callouts/python/extproc/service/network_callout_server.py

```diff
@@ -0,0 +1,364 @@
+    if self._health_check_server:
+      self._health_check_server.shutdown()
+    if self._callout_server:
+      self._callout_server.stop()
+
+  def process(
+      self,
+      callout: ProcessingRequest,
+      context: ServicerContext,
+  ) -> ProcessingResponse:
+    """Process incoming network callouts.
+
+    Args:
+        callout: The incoming network callout.
+        context: Stream context on the callout.
+
+    Returns:
+        ProcessingResponse: A response for the incoming callout.
+    """
+    response = ProcessingResponse()
+
+    # Handle read data (client->server)
+    if callout.HasField('read_data'):
+      data = callout.read_data.data
+      end_of_stream = callout.read_data.end_of_stream
+
+      # Process the data through callback
+      processed_data, modified = self.on_read_data(data, end_of_stream, context)
+      
+      # Build response
+      response.read_data.data = processed_data
+      response.read_data.end_of_stream = end_of_stream
+      response.data_processing_status = (
+          ProcessingResponse.MODIFIED if modified
+          else ProcessingResponse.UNMODIFIED
+      )
+
+      # Check connection control
+      conn_status = self.should_close_connection(data, modified, context)
+
+    # Handle write data (server->client)
+    elif callout.HasField('write_data'):
+      data = callout.write_data.data
+      end_of_stream = callout.write_data.end_of_stream
+
+      # Process the data through callback
+      processed_data, modified = self.on_write_data(data, end_of_stream, context)
+
+      # Build response
+      response.write_data.data = processed_data
+      response.write_data.end_of_stream = end_of_stream
+      response.data_processing_status = (
+          ProcessingResponse.MODIFIED if modified
+          else ProcessingResponse.UNMODIFIED
+      )
+
+      # Check connection control
+      conn_status = self.should_close_connection(data, modified, context)
+
+    else:
+      logging.warning("Received request with no data")
+      return response
+
+    # Set connection status
+    response.connection_status = (
+        ProcessingResponse.CLOSE if conn_status
+        else ProcessingResponse.CONTINUE
+    )
+
+    return response
+
+  def on_read_data(
+      self,
+      data: bytes,
+      end_of_stream: bool,
+      context: ServicerContext,
+  ) -> Tuple[bytes, bool]:
+    """Process data from client to server (read path).
+    
+    Override this method to implement custom processing logic.
+    
+    Args:
+        data: Raw bytes from the client
+        end_of_stream: Whether this is the last data frame
+        context: gRPC context
+        
+    Returns:
+        Tuple of (processed_data, modified)
+    """
+    # Default: pass through unchanged
+    return data, False
+
+  def on_write_data(
+      self,
+      data: bytes,
+      end_of_stream: bool,
+      context: ServicerContext,
+  ) -> Tuple[bytes, bool]:
```

Comment:
Same comment as for on_read_data: consider using a namedtuple or dataclass.

Reply:
done.